go-controller/pkg/ovn/ovn.go

```diff
@@ -59,23 +61,63 @@ type ACLLoggingLevels struct {
 	Deny  string `json:"deny,omitempty"`
 }
 
-// Controller structure is the object which holds the controls for starting
-// and reacting upon the watched resources (e.g. pods, endpoints)
-type Controller struct {
+// GenericController structure is place holder for all fields shared among controllers.
+type GenericController struct {
 	client       clientset.Interface
 	kube         kube.Interface
 	watchFactory *factory.WatchFactory
-	stopChan     <-chan struct{}
+	podRecorder  metrics.PodRecorder
 
-	// FIXME DUAL-STACK -  Make IP Allocators more dual-stack friendly
-	masterSubnetAllocator        *subnetallocator.SubnetAllocator
-	hybridOverlaySubnetAllocator *subnetallocator.SubnetAllocator
+	// event recorder used to post events to k8s
+	recorder record.EventRecorder
+
+	// libovsdb northbound client interface
+	nbClient libovsdbclient.Client
+
+	// libovsdb southbound client interface
+	sbClient libovsdbclient.Client
 
+	// has SCTP support
 	SCTPSupport bool
+}
+
+// ControllerManager structure is the object manages all controllers for all networks
+type ControllerManager struct {
+	GenericController
+
+	// default wait group and stop channel, used by default network controller
+	defaultWg       *sync.WaitGroup
+	defaultStopChan <-chan struct{}
+
+	// unique identity for controllerManager running on different ovnkube-master instance,
+	// used for leader election
+	identity string
+
+	// default network controller
+	defaultController *Controller
```

Comment:
I  would expect ControllerManager to manage the default OvnController almost in the same way as it manages other OvnController(s), so having specific attributes for the default one throws me off. But I can't see enough on this PR to tell where this is heading.

Reply:
I looked through our current code, the reference of this defaultController field is limited to initialization.

Reply:
But if it is not read after initialization, what's the use of it?

Reply:
because initialization code need it.

Reply:
It just looks like everything regarding the default controller should be managed within `InitDefaultController` and that we could just look into the map for the default controller instead of having a specific variable.

```
func (cm *ControllerManager) InitDefaultController(addressSetFactory addressset.AddressSetFactory) (*Controller, error) {
  if c, ok := allOvnControllers["default"]; ok {
     return c, nil
  }
  
  // init default controller

  allOvnControllers["default"]=c
  rerturn c, nil

}
```

The defaultController variable seems to be read and written all over the place instead of having the classic lazy initialization method for it and feels weird.

Reply:
In general, when ovnkube-master starts, a controller (either default or secondary) can only be started when all existing net-attach-defs are handled, only when, all nads of the that controller is known (saved in the controller), and watchPods() would be able to determine if interface belongs to this controller needs to be plumbed.
   
that is why we moved defaultController.Start() afterwards.

Reply:
> In general, when ovnkube-master starts, a controller (either default or secondary) can only be started when all existing net-attach-defs are handled, only when, all nads of the that controller is known (saved in the controller), and watchPods() would be able to determine if interface belongs to this controller needs to be plumbed.
> 
> that is why we moved defaultController.Start() afterwards.

NADs can be added any time in the lifetime of the system.

So how it is going to be decided that all NADs for a network are known?
Conversely, how is the user going to be made aware if a NAD has been considered for a network?

Also, this does not apply for the default controller/network as the relationship is 1:1 and everything about it is well-known at initialization.

This is confusing because I am not sure if you are answering to this comment or some other one about initialization, can you clarify?

Reply:
yes, nad can be added any time. But I am referring to the case when ovnkube-master restarts, and only for all existing default nads need to be handled before defaultControler.Start()->watchPods(), ad the existing pods may request any exsting defaut nad to be plumbed.

No, there is no 1:1 relationship to the controller/nad. multiple nads can be specified for the same network (controller). Although they must share the same netconf Name, and netcidr/mtu etc., but they have different namespace/name, and may have different resources (for example, request different types of sriov resources).

Reply:
> for all existing default nads

What are the default nads? As far as I know there is only one default nad for the default primary network.

> @jcaamano yes, nad can be added any time. But I am referring to the case when ovnkube-master restarts, and only for all existing default nads (not nads added after that), those need to be handled before defaultControler.Start()->watchPods(), ad the existing pods may request any exsting defaut nad to be plumbed.

I don't really understand how is that any different. A nad can be added at the same ovnkube-master (re)starts just as it can be added at the same time a pod consuming that nad is added. It is the same problem of deciding when it is considered that all NADs for a network are known; and I wouldn't think we need different solutions for that problem whether we are initializing or already running.

As I said, I am not sure this is related to the default controller initialization I talked about in the comment you are answering to?

> No, there is no 1:1 relationship to the controller/nad. multiple nads can be specified for the same network (controller). Although they must share the same netconf Name, and netcidr/mtu etc., but they have different namespace/name, and may have different resources (for example, request different types of sriov resources).

I am talking about the default primary network for which there will only ever be 1 (fictitious) nad, right? So we don't need to know about any other nads to initialize the controller for the default primary network.

In any case, I need to take a look at the newest push.